admittance_controller/include/admittance_controller/admittance_controller.hpp

```diff
@@ -0,0 +1,148 @@
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    CallbackReturn on_cleanup(const rclcpp_lifecycle::State &previous_state) override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    CallbackReturn on_error(const rclcpp_lifecycle::State &previous_state) override;
+
+    ADMITTANCE_CONTROLLER_PUBLIC
+    controller_interface::return_type update_and_write_commands(
+        const rclcpp::Time &time, const rclcpp::Duration &period) override;
+
+
+  protected:
+    std::vector<hardware_interface::CommandInterface> on_export_reference_interfaces() override;
+
+    controller_interface::return_type update_reference_from_subscribers() override;
+
+    bool on_set_chained_mode(bool chained_mode) override;
+
+    int num_joints_{};
+
+    std::vector<std::reference_wrapper<hardware_interface::LoanedCommandInterface>> joint_position_command_interface_;
+    std::vector<std::reference_wrapper<hardware_interface::LoanedCommandInterface>> joint_velocity_command_interface_;
+    std::vector<std::reference_wrapper<hardware_interface::LoanedCommandInterface>> joint_acceleration_command_interface_;
+    std::vector<std::reference_wrapper<hardware_interface::LoanedCommandInterface>> joint_effort_command_interface_;
+    std::vector<std::reference_wrapper<hardware_interface::LoanedStateInterface>> joint_position_state_interface_;
+    std::vector<std::reference_wrapper<hardware_interface::LoanedStateInterface>> joint_velocity_state_interface_;
+    std::vector<std::reference_wrapper<hardware_interface::LoanedStateInterface>> joint_acceleration_state_interface_;
+
+    std::vector<double *> position_reference_;
+    std::vector<double *> velocity_reference_;
+
+    // Admittance rule and dependent variables;
+    std::unique_ptr<admittance_controller::AdmittanceRule> admittance_;
+    // joint limiter TODO
+    std::unique_ptr<semantic_components::ForceTorqueSensor> force_torque_sensor_;
+    // ROS subscribers
+    rclcpp::Subscription<trajectory_msgs::msg::JointTrajectoryPoint>::SharedPtr input_joint_command_subscriber_ = nullptr;
+    rclcpp::Publisher<control_msgs::msg::AdmittanceControllerState>::SharedPtr s_publisher_ = nullptr;
+    // ROS messages
+    std::shared_ptr<trajectory_msgs::msg::JointTrajectoryPoint> joint_command_msg;
+    // real-time buffer
+    RTBuffers rtBuffers;
+
+    // controller running state
+    bool controller_is_active_{};
+    const std::set<std::string> allowed_state_interface_types_ = {
+        hardware_interface::HW_IF_POSITION,
+        hardware_interface::HW_IF_VELOCITY,
+    };
+    trajectory_msgs::msg::JointTrajectoryPoint last_commanded_state_;
+    trajectory_msgs::msg::JointTrajectoryPoint last_state_reference_;
+    trajectory_msgs::msg::JointTrajectoryPoint state_offset_;
+    trajectory_msgs::msg::JointTrajectoryPoint prev_trajectory_point_;
+    // control loop data
+    trajectory_msgs::msg::JointTrajectoryPoint state_reference_, state_current_, state_desired_,
```

Comment:
I don't mind it because it make it more clear that it contains the entire state (position, velocity, etc.) instead of just position. Maybe it would read better as `reference_state_`,  `current_state_`, and `desired_state_`?